### PR TITLE
Fix missing header, refactor with Threads::DisablePerfLogInScope

### PR DESCRIPTION
**Pull request by roystgnr**
In poly2tri_triangulator.C we were getting libmesh_logging.h indirectly from threads_foo.h of all places, in a way that broke if neither of TBB and pthreads were available.  See #3659.

After fixing that, I figured I'd avoid the problem happening again, by getting that header out of the threads headers and into threads.C ... which turns out to be a cleaner and more exception-safe solution too.

We should wait on merging this until Logan's done with the CI backend transition.

### CODE REVIEW

File: include/parallel/threads.h
Comment (jwpeterson) [inline include/parallel/threads.h:74]: `= default;` for the "non logging" implementations rather than empty brackets? 

The [reasoning](https://stackoverflow.com/questions/13576055/how-is-default-different-from-for-default-constructor-and-destructor) is related to whether the compiler can consider it a "trivial" class or not.

Comment (roystgnr) [reply include/parallel/threads.h:74]: Oh, man, I'd forgotten about that.  Will fix.

**Comment by roystgnr** [comment]
"Unused variable"?  Oh, of course.  If we build with --disable-perflog then the constructor and destructor do nothing, and g++ is smart enough to recognize that.  Clever.

**Comment by roystgnr** [comment]
Pinging @loganharbour in case he wants to look at the coverage diff.  A bunch of new misses, but all on `continue;` keywords or `[this]` lambda captures, so it looks to me like the new Civet environment is just optimizing a little differently, not missing any configuration details that would eat into our test coverage.

**Comment by loganharbour** [comment]
> Pinging @loganharbour in case he wants to look at the coverage diff. A bunch of new misses, but all on `continue;` keywords or `[this]` lambda captures, so it looks to me like the new Civet environment is just optimizing a little differently, not missing any configuration details that would eat into our test coverage.

Yeah, this looks fine to me. I thought that the new base would already be updated, but looks like this one has a different base. Good to go

**Comment by roystgnr** [comment]
Thanks!  I'll wait for John to have a chance to review if he wants; last I talked to him I was still pushing "let's not merge anything until the Civet updates are done".

In fact ... do we have any Civet coverage using TBB?  I used to use that heavily, but looking at my own local builds, the newest thing I can find that isn't instead relying on the pthread backend is from 2021... and looking at the LCOV results here I see `threads_pthread.h` but not `threads_tbb.h`, so that's a no.  I'll see if I can test this locally, explicitly with TBB, and make sure I haven't broken that.

**Review by jwpeterson — APPROVED** [review]
Sorry, meant to approve this when I commented before.

I think this is the first time we've used the `[[maybe_unused]]` attribute, so ideally we should have a configure test for that, even though it *should* be supported by all C++17 compliant compilers.

**Comment by roystgnr** [comment]
> ideally we should have a configure test for that

I think we do!  It's in the test body used by ax_cxx_compile_stdcxx.m4

I'm not sure how to *test* the test, though.  Wrap a compliant C++17 compiler in a shell script that first greps for maybe_unused and dies if it finds it?

**Comment by jwpeterson** [comment]
> I think we do! It's in the test body used by ax_cxx_compile_stdcxx.m4

OK, thanks for pointing that out. I didn't find it with a cursory search since it's actually in the autoconf-submodule and not libmesh's m4 directory. That's definitely good enough for me.

**Pull request merged**
